File: solutions/typescript/resistor-color-trio/1/resistor-color-trio.ts
const colorSet = new Map<string, number>([
  ["black", 0],
  ["brown", 1],
  ["red", 2],
  ["orange", 3],
  ["yellow", 4],
  ["green", 5],
  ["blue", 6],
  ["violet", 7],
  ["grey", 8],
  ["white", 9],
]);

export function decodedResistorValue(colors: string[]): string {
  const colorDuo = decodedValue(colors);
  return getMetricPrefix(colorDuo, colors);
}

function decodedValue(colors: string[]): number {
  let val = 0;
  if (colors.length < 1)
    throw new Error("Array must not be empty");

  if (colorSet.has(colors[0]))
    val += colorSet.get(colors[0])!;

  if (colors.length > 1 && colorSet.has(colors[1])) {
    val *= 10;
    val += colorSet.get(colors[1])!;
  }
  return val;
}

function getMetricPrefix(num: number, colors: string[]): string {
  const giga = 1_000_000_000;
  const mega = 1_000_000;
  const kilo = 1_000;
  const exponent = colorSet.get(colors[2]) ?? 0;
  for (let i = exponent; i > 0; i--)
    num *= 10;

  if(num >= giga)
    return `${getBaseNumber(num, giga)} gigaohms`;
  if (num >= mega)
    return `${getBaseNumber(num, mega)} megaohms`;
  if(num >= kilo){
    return `${getBaseNumber(num, kilo)} kiloohms`;
  }
  return `${num} ohms`;
}

function getBaseNumber(num: number, baseNum: number): number {
  while(num % baseNum === 0)
    num /= baseNum;
  return num;
}

File: solutions/csharp/resistor-color-trio/1/ResistorColorTrio.cs
public static class ResistorColorTrio
{
    private static Dictionary<string, int> Colors = new Dictionary<string, int>()
    {
        { "black", 0 },
        { "brown", 1 },
        { "red", 2 },
        { "orange", 3 },
        { "yellow", 4 },
        { "green", 5 },
        { "blue", 6 },
        { "violet", 7 },
        { "grey", 8 },
        { "white", 9 },
    };
    private const int Giga = 1_000_000_000;
    private const int Mega = 1_000_000;
    private const int Kilo = 1_000;

    public static string Label(string[] colors)
    {
        string[] lowerCaseColors = colors.Select(c => c.ToLower()).ToArray();
        long decodedValue = GetDecodedValue(lowerCaseColors);
        return GetDecodedOutput(decodedValue, lowerCaseColors);
    }

    private static long GetDecodedValue(string[] colors)
    {
        long val = 0;
        if (colors.Length < 1)
            throw new ArgumentOutOfRangeException();

        if (Colors.ContainsKey(colors[0]))
            val += Colors[colors[0]];

        if (Colors.Count > 1 && Colors.ContainsKey(colors[1]))
        {
            val *= 10;
            val += Colors[colors[1]];
        }
        if (Colors.Count > 2 && Colors.ContainsKey(colors[2]))
        {
            int exponent = Colors[colors[2]];
            for (int i = exponent; i > 0; i--)
                val *= 10;
        }
        return val;
    }

    private static string GetDecodedOutput(long decodedValue, string[] colors)
    {
        if (decodedValue >= Giga)
            return $"{GetBaseNumber(decodedValue, Giga)} gigaohms";
        if (decodedValue >= Mega)
            return $"{GetBaseNumber(decodedValue, Mega)} megaohms";
        if (decodedValue >= Kilo)
            return $"{GetBaseNumber(decodedValue, Kilo)} kiloohms";
        return $"{decodedValue} ohms";
    }

    private static long GetBaseNumber(long decodedValue, int baseNumber)
    {
        while (decodedValue % baseNumber == 0)
            decodedValue /= baseNumber;
        return decodedValue;
    }
}
